src/main/lrsql/admin/interceptors/oidc.clj

```diff
@@ -0,0 +1,90 @@
+    :enter
+    (fn validate-oidc-identity [{lrs :com.yetanalytics/lrs
+                                 :as ctx}]
+      (if-let [ret (oidc/token-auth-admin-identity
+                    ctx
+                    (get-in lrs [:config :oidc-scope-prefix]))]
+        (if (= ::oidc/unauthorized ret)
+          (assoc (chain/terminate ctx)
+                 :response
+                 {:status 401
+                  :body   {:error "Invalid JWT Claims!"}})
+          (assoc ctx ::admin-identity ret))
+        ;; No OIDC
+        ctx))}))
+
+;; NOTE: Right now there is only one blanket :scope/admin
+;; Currently only applied when OIDC is used.
+(def authorize-oidc-request
+  "If an admin identity is present, check if it has the proper scope(s) for the
+  given action. No-op with no identity."
+  (interceptor
+   {:name ::authorize-oidc-request
+    :enter
+    (fn authorize-oidc-request [{admin-identity ::admin-identity
+                                 :as            ctx}]
+      (if admin-identity
+        (if (auth/authorize-action ctx admin-identity)
+          ctx
+          (assoc (chain/terminate ctx)
+                 :response
+                 {:status 401
+                  :body   {:error "Unauthorized Admin Action!"}}))
+        ctx))}))
+
+(defn- disable-jwt-interceptors
+  [{queue ::chain/queue :as ctx}]
+  (assoc ctx
+         ::chain/queue
+         (into (empty queue)
+               (remove
+                (comp
+                 #{::jwt/validate-jwt
+                   ::jwt/validate-jwt-account}
+                 :name)
+                queue))))
+
+(def ensure-oidc-identity
+  "If an admin identity is present, create or return the user, validating the
+  issuer. On success, inject the account ID and disable subsequent JWT
+  interceptors. No-op with no identity."
+  (interceptor
+   {:name ::ensure-oidc-identity
+    :enter
+    (fn ensure-oidc-identity [{lrs            :com.yetanalytics/lrs
+                               admin-identity ::admin-identity
+                               :as            ctx}]
+      (if admin-identity
+        (let [{:keys [username
+                      oidc-issuer]} admin-identity
+              {:keys [result]} (adp/-ensure-account-oidc
+                                lrs
+                                username
+                                oidc-issuer)]
+          (if (= :lrsql.admin/oidc-issuer-mismatch-error result)
+            (assoc (chain/terminate ctx)
+                   :response
+                   {:status 401
+                    :body   {:error "OIDC Issuer Mismatch!"}})
+            (-> ctx
+                disable-jwt-interceptors
+                (assoc-in [::jwt/data :account-id] result)
+                ;; NOTE: I do this to follow what is done for native JWT
+                ;; Don't think this app actually uses a session at all?
+                (assoc-in
+                 [:request :session ::jwt/data :account-id]
+                 result))))
```

Comment:
Was curious about this, as I found several places where session values are set, but none where they get used. Additionally no kind of session config or store is present in the service map. When I see it in the jwt interceptors, is it vestigial?

Reply:
I believe I added that code so that the account-id data is in two places (the JWT as well as higher up in the Pedestal context) as a form of redundancy. I guess it's not strictly needed, but I think it's good to have that extra info in the JWT.

Reply:
Sure, I have no problem with it being there, it just reads like there is going to be some session to hold that data (per ring conventions). That said, you will often see duplicate data placed on the context _and_ the request, this is generally to provide it to function-style terminal interceptors ("handlers"), that only ever get the request.